pyvista/plotting/render_passes.py

```diff
@@ -0,0 +1,255 @@
+    'vtkDepthOfFieldPass',
+    'vtkGaussianBlurPass',
+    'vtkOpenGLFXAAPass',
+    'vtkSSAAPass',  # should really be last
+]
+
+
+class RenderPasses:
+    """Class to support multiple render passes for a renderer.
+
+    Notes
+    -----
+    Passes are organized here as "Primary" (vtkOpenGLRenderPass) that act
+    within the renderer and "post-processing" (vtkImageProcessingPass) passes,
+    which act on the image generated from the renderer.
+
+    The primary passes are added as part of a vtk.vtkRenderPassCollection or
+    are "stacked", while the post-processing passes are added as a final pass
+    to the rendered image.
+
+    """
+
+    def __init__(self, renderer):
+        """Initialize render passes."""
+        self._renderer_ref = weakref.ref(renderer)
+
+        self._passes = {}
+        self._fxaa_pass = None
+        self._shadow_map_pass = None
+        self._edl_pass = None
+        self._dof_pass = None
+        self._ssaa_pass = None
+        self._blur_passes = []
+        self.__pass_collection = None
+        self.__seq_pass = None
+        self.__camera_pass = None
+
+    @property
+    def _pass_collection(self):
+        """Initialize (when necessary the pass collection and return it.
+
+        This lets us lazily generate the pass collection only when we need it
+        rather than at initialization of the class.
+
+        """
+        if self.__pass_collection is None:
+            self._init_passes()
+        return self.__pass_collection
+
+    @property
+    def _seq_pass(self):
+        """Initialize (when necessary the pass collection and return it.
+
+        This lets us lazily generate the pass collection only when we need it
+        rather than at initialization of the class.
+
+        """
+        if self.__seq_pass is None:
+            self._init_passes()
+        return self.__seq_pass
+
+    @property
+    def _camera_pass(self):
+        """Initialize (when necessary the pass collection and return it.
+
+        This lets us lazily generate the pass collection only when we need it
+        rather than at initialization of the class.
+
+        """
+        if self.__camera_pass is None:
+            self._init_passes()
+        return self.__camera_pass
+
+    def _init_passes(self):
+        """Initialize the renderer's standard passes."""
+        # simulate the standard VTK rendering passes and put them in a sequence
+        self.__pass_collection = _vtk.vtkRenderPassCollection()
+        self.__pass_collection.AddItem(_vtk.vtkRenderStepsPass())
+
+        self.__seq_pass = _vtk.vtkSequencePass()
+        self.__seq_pass.SetPasses(self._pass_collection)
+
+        # Make the sequence the delegate of a camera pass.
+        self.__camera_pass = _vtk.vtkCameraPass()
+        self.__camera_pass.SetDelegatePass(self._seq_pass)
+
+    @property
+    def _renderer(self):
+        """Return the renderer."""
+        if self._renderer_ref is not None:
+            return self._renderer_ref()
```

Comment:
Was this meant to say
```suggestion
        if self._renderer_ref() is not None:
            return self._renderer_ref()
```
?

Reply:
either way, why even have the check? it'll return `None` if the condition evaluates to false (being `None`)

```suggestion
        return self._renderer_ref()
```

Reply:
Well if the original version is correct, then it's checking for the case of `self._renderer_ref = None` (rather than a stale weakref), in which case trying to call `self._renderer_ref()` would give a `TypeError`.